modules/extensions/src/collide/collide-effect.ts

```diff
@@ -0,0 +1,252 @@
+    {
+      effects: allEffects,
+      layers,
+      layerFilter,
+      viewports,
+      onViewportActive,
+      views,
+      pass,
+      preRenderStats = {}
+    }: PreRenderOptions
+  ): void {
+    if (!this.dummyCollideMap) {
+      this.dummyCollideMap = new Texture2D(gl, {width: 1, height: 1});
+    }
+
+    if (pass.startsWith('picking')) {
+      // Do not update on picking pass
+      return;
+    }
+
+    const collideLayers = layers.filter(
+      // @ts-ignore
+      ({props: {visible, collideEnabled}}) => visible && collideEnabled
+    ) as Layer<CollideExtensionProps>[];
+    if (collideLayers.length === 0) {
+      this.channels = {};
+      return;
+    }
+
+    // Detect if mask has rendered. TODO: better dependency system for Effects
+    const effects = allEffects?.filter(e => e.constructor === MaskEffect);
+    const maskEffectRendered = (preRenderStats['mask-effect'] as MaskPreRenderStats)?.didRender;
+
+    // Collect layers to render
+    const channels = this._groupByCollideGroup(gl, collideLayers);
+
+    const viewport = viewports[0];
+    const viewportChanged =
+      !this.lastViewport || !this.lastViewport.equals(viewport) || maskEffectRendered;
+
+    // Resize framebuffers to match canvas
+    for (const collideGroup in channels) {
+      const collideFBO = this.collideFBOs[collideGroup];
+      const renderInfo = channels[collideGroup];
+      collideFBO.resize({
+        width: gl.canvas.width / DOWNSCALE,
+        height: gl.canvas.height / DOWNSCALE
+      });
+      this._render(renderInfo, {
+        effects,
+        layerFilter,
+        onViewportActive,
+        views,
+        viewport,
+        viewportChanged
+      });
+    }
+
+    // debugFBO(this.collideFBOs[Object.keys(channels)[0]], {minimap: true});
+  }
+
+  private _render(
+    renderInfo: RenderInfo,
+    {
+      effects,
+      layerFilter,
+      onViewportActive,
+      views,
+      viewport,
+      viewportChanged
+    }: {
+      effects: PreRenderOptions['effects'];
+      layerFilter: PreRenderOptions['layerFilter'];
+      onViewportActive: PreRenderOptions['onViewportActive'];
+      views: PreRenderOptions['views'];
+      viewport: Viewport;
+      viewportChanged: boolean;
+    }
+  ) {
+    const {collideGroup} = renderInfo;
+    const oldRenderInfo = this.channels[collideGroup];
+    if (!oldRenderInfo) {
+      return;
+    }
+
+    const renderInfoUpdated =
+      // If render info is new
+      renderInfo === oldRenderInfo ||
+      // If sublayers have changed
+      oldRenderInfo.layers.length !== renderInfo.layers.length ||
+      // If a sublayer's bounds have been updated
+      renderInfo.layerBounds.some((b, i) => !equals(b, oldRenderInfo.layerBounds[i]));
+
+    this.channels[collideGroup] = renderInfo;
+
+    if (renderInfoUpdated || viewportChanged) {
+      this.lastViewport = viewport;
+      const collideFBO = this.collideFBOs[collideGroup];
+      const collidePass = this.collidePasses[collideGroup];
```

Comment:
Is there any reason `CollidePass` cannot be reused across multiple channels?

Reply:
Good point, just using one now